File: android/src/main/kotlin/com/yueying/network_status_bridge/NetworkStatusBridgePlugin.kt
package com.yueying.network_status_bridge

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/** NetworkStatusBridgePlugin */
class NetworkStatusBridgePlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    private var eventToken: String? = null
    private val scope = MainScope()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel =
            MethodChannel(flutterPluginBinding.binaryMessenger, "network_status_bridge/method")
        methodChannel.setMethodCallHandler(this)
        //
        eventChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, "network_status_bridge/event")
        eventChannel.setStreamHandler(this)
        //
        flutterPluginBinding.applicationContext.let(NetworkMonitor::startMonitoring)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        scope.cancel()
    }


    override fun onMethodCall(call: MethodCall, result: Result) = when (call.method) {
        "getCurrentType" ->
            result.success(NetworkMonitor.type.value)

        else -> result.notImplemented()
    }


    override fun onListen(arguments: Any?, events: EventSink?) {
        // 先注册监听，确保不丢失任何变化
        eventToken = NetworkMonitor.addObserver {
            scope.launch {
                events?.success(it.value)
            }
        }
        // 再发送当前网络状态作为初始值（注册后读取，保证是最新的）
        events?.success(NetworkMonitor.type.value)
    }

    override fun onCancel(arguments: Any?) {
        eventToken?.let(NetworkMonitor::removeObserver)
        eventToken = null
    }
}

File: ios/Classes/NetworkStatusBridgePlugin.swift
import Flutter

public final class NetworkStatusBridgePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    
    private var eventSink: FlutterEventSink?
    private var token: String?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        let methodChannel =  FlutterMethodChannel(name: "network_status_bridge/method", binaryMessenger: messenger)
        let eventChannel = FlutterEventChannel(name: "network_status_bridge/event", binaryMessenger: messenger)
        //
        let instance = NetworkStatusBridgePlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
        //
        YYINetworkMonitor.shared.startMonitoring()
    }
    
    ///
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getCurrentType":
            result(YYINetworkMonitor.shared.currentType.rawValue)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    ///
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        token = YYINetworkMonitor.shared.addObserver(networkObserver)
        return nil;
    }
    
    ///
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        if let token = token {
            YYINetworkMonitor.shared.removeObserver(token)
            self.token = nil
        }
        eventSink = nil
        return nil;
    }
    
    ///
    private func networkObserver(_ type: YYINetworkType) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(type.rawValue)
        }
    }
}
